docs/tutorials/nextjs.mdx

```diff
@@ -0,0 +1,186 @@
+export type RootState = ReturnType<StoreType['getState']>
+// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
+export type AppDispatch = StoreType['dispatch']
+
+// file: app/StoreProvider.tsx
+;('use client')
+import { useRef } from 'react'
+import { Provider } from 'react-redux'
+// highlight-start
+import { createStore } from './store'
+// highlight-end
+
+export default function StoreProvider({
+  children,
+}: {
+  children: React.ReactNode
+}) {
+  // highlight-start
+  const storeRef = useRef<ReturnType<typeof createStore>>(createStore())
+  // highlight-end
+
+  return <Provider store={storeRef.current}>{children}</Provider>
+}
+```
+
+If you need to intialize the store with data from the parent component then define that data as a property and use an action on the slice to set the data in the store as shown below.
+
+```ts title="src/app/StoreProvider.tsx"
+// file: features/counter/counterSlice.ts noEmit
+import { createSlice } from '@reduxjs/toolkit'
+import type { PayloadAction } from '@reduxjs/toolkit'
+
+const counterSlice = createSlice({
+  name: 'counter',
+  initialState: {
+    value: 0,
+  },
+  reducers: {
+    setCount: (state, action: PayloadAction<number>) => {
+      state.value = action.payload
+    },
+  },
+})
+
+export const { setCount } = counterSlice.actions
+export default counterSlice.reducer
+
+// file: app/store.ts noEmit
+import { configureStore } from '@reduxjs/toolkit'
+import counterReducer from '../features/counter/counterSlice'
+
+export const createStore = () =>
+  configureStore({
+    reducer: {
+      counter: counterReducer,
+    },
+  })
+
+// Infer the type of createStore
+export type StoreType = ReturnType<typeof createStore>
+// Infer the `RootState` and `AppDispatch` types from the store itself
+export type RootState = ReturnType<StoreType['getState']>
+// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
+export type AppDispatch = StoreType['dispatch']
+
+// file: app/StoreProvider.tsx
+;('use client')
+import { useRef } from 'react'
+import { Provider } from 'react-redux'
+import { createStore } from './store'
+// highlight-start
+import { setCount } from '../features/counter/counterSlice'
+// highlight-end
+
+export default function StoreProvider({
+  count,
+  children,
+}: {
+  count: number
+  children: React.ReactNode
+}) {
+  const storeRef = useRef<ReturnType<typeof createStore> | null>(null)
+  if (!storeRef.current) {
+    storeRef.current = createStore()
+    // highlight-start
+    storeRef.current.dispatch(setCount(count))
+    // highlight-end
+  }
+
+  return <Provider store={storeRef.current}>{children}</Provider>
+}
+```
+
+The next step is to include the `StoreProvider` in your layout component. This will ensure that the store is created for each request and that the store is not shared across requests. Use the store exactly as you would normally using the hooks provided by `react-redux`.
```

Comment:
```suggestion
The next step is to include the `StoreProvider` in your layout component. This will ensure that the store is created for each request and that the store is not shared across requests. In all client components further down the tree, you can use the store exactly as you would normally using the hooks provided by `react-redux`.
```